backend-lib/src/main/rust/lib.rs

```diff
@@ -290,6 +290,74 @@ pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_initDataD
     unwrap_exc_or(&mut env, res, -1)
 }
 
+const JNI_ACCOUNT: &str = "cash/z/ecc/android/sdk/internal/model/JniAccount";
+
+fn encode_account<'a, P: Parameters>(
+    env: &mut JNIEnv<'a>,
+    network: &P,
+    account: zcash_client_sqlite::wallet::Account,
+) -> jni::errors::Result<JObject<'a>> {
+    let ufvk = match account.ufvk() {
+        Some(ufvk) => env.new_string(ufvk.encode(network))?.into(),
+        None => JObject::null(),
+    };
+
+    env.new_object(
+        JNI_ACCOUNT,
+        "(JLjava/lang/String;)V",
+        &[
+            // TODO: This will be replaced by the multi-seed-compatible account ID.
+            JValue::Long(i64::from(account.id().as_u32())),
+            (&ufvk).into(),
+        ],
+    )
+}
+
+/// Fetches the accounts in the given wallet.
+#[unsafe(no_mangle)]
+pub extern "C" fn Java_cash_z_ecc_android_sdk_internal_jni_RustBackend_getAccounts<'local>(
+    mut env: JNIEnv<'local>,
+    _: JClass<'local>,
+    db_data: JString<'local>,
+    network_id: jint,
+) -> jobjectArray {
+    let res = catch_unwind(&mut env, |env| {
+        let network = parse_network(network_id as u32)?;
+        let db_data = wallet_db(env, network, db_data)?;
+
+        let accounts = db_data
+            .get_account_ids()?
+            .into_iter()
+            .map(|account_id| {
+                db_data
+                    .get_account(account_id)
+                    .transpose()
+                    .expect("account_id exists")
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let first_account = accounts.first().cloned();
+
+        Ok(utils::rust_vec_to_java(
+            env,
+            accounts,
+            JNI_ACCOUNT,
+            |env, account| encode_account(env, &network, account),
+            |env| {
+                // The array contains non-null values in Kotlin. This returns `null` if
+                // there are no accounts, which is fine because the array has no entries
+                // and thus the empty element is never used.
+                match first_account {
+                    Some(account) => encode_account(env, &network, account),
+                    None => Ok(JObject::null()),
+                }
+            },
```

Comment:
Doesn't this end up calling `encode_account` twice redundantly, when `first_account` is Some? (Once to create the `jempty` object that the array is initialized to; again to set the only element.) It might be better to have a `utils::rust_option_to_java(` utility that maps None to an empty Java array and Some to a single-element array. But maybe it's not worth it because this is going to change anyway?

https://github.com/Electric-Coin-Company/zcash-android-wallet-sdk/blob/e9a822fc50e4df8ef79651a52b35b9910038d8db/backend-lib/src/main/rust/utils.rs#L54-L74

Reply:
I figured out how to do this more elegantly in the general case without needing `empty_element` at all. Filed #1641.